sympy/solvers/ode/ode.py

```diff
@@ -2937,7 +2955,14 @@ def _constant_renumber(expr):
     constants_found = [c for c in constants_found if c not in variables]
 
     # Renumbering happens here
-    expr = expr.subs(zip(constants_found[1:], iter_constants), simultaneous=True)
+
+    expr_subs = expr if isinstance(expr, (set, list, tuple)) else [expr]
+
+    subs_dict = {var: cons for var, cons in zip(constants_found[1:-1], iter_constants)}
+    expr_subs = [e.subs(subs_dict, simultaneous=True)
+            for e in expr_subs]
+    expr = type(expr)(expr_subs) if isinstance(expr, (set, list, tuple)) else expr_subs[0]
```

Comment:
When you find yourself checking `if isinstance` repeatedly it's a clear sign that there is probably a better way to organise this. At the beginning of the function you could do
```
if isinstance(expr, (set, list, tuple)):
    return type(expr)(constant_renumber(Tuple(*expr)))
```
A function where you don't know whether or not something is `Expr` or `list[Expr` etc is a mess so you should handle that case right at the beginning.